src/data_cell/graph_datacell.h

```diff
@@ -0,0 +1,155 @@
+
+// Copyright 2024-present the vsag project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <limits>
+#include <memory>
+#include <nlohmann/json.hpp>
+#include <unordered_map>
+#include <vector>
+
+#include "algorithm/hnswlib/hnswalg.h"
+#include "common.h"
+#include "graph_interface.h"
+#include "index/index_common_param.h"
+#include "io/basic_io.h"
+#include "vsag/constants.h"
+
+namespace vsag {
+
+/**
+ * built by nn-descent or incremental insertion
+ * add neighbors and pruning
+ * retrieve neighbors
+ */
+template <typename IOTmpl, bool is_adapter>
+class GraphDataCell;
+
+template <typename IOTmpl>
+class GraphDataCell<IOTmpl, false> : public GraphInterface {
+public:
+    GraphDataCell(const nlohmann::json& graph_json_params,
+                  const nlohmann::json& io_json_params,
+                  const IndexCommonParam& common_param);
+
+    void
+    InsertNeighborsById(InnerIdType id, const Vector<InnerIdType>& neighbor_ids) override;
+
+    uint32_t
+    GetNeighborSize(InnerIdType id) override;
+
+    void
+    GetNeighbors(InnerIdType id, Vector<InnerIdType>& neighbor_ids) override;
+
+    inline void
+    SetIO(std::shared_ptr<BasicIO<IOTmpl>> io) {
+        this->io_ = io;
+    }
+
+    /****
+     * prefetch neighbors of a base point with id
+     * @param id of base point
+     * @param neighbor_i index of neighbor, 0 for neighbor size, 1 for first neighbor
+     */
+    void
+    Prefetch(InnerIdType id, uint32_t neighbor_i) override {
+        io_->Prefetch(id * this->code_line_size_ + sizeof(uint32_t) +
+                      neighbor_i * sizeof(InnerIdType));
+    }
+
+    void
+    Serialize(StreamWriter& writer) override;
+
+    void
+    Deserialize(StreamReader& reader) override;
+
+private:
+    std::shared_ptr<BasicIO<IOTmpl>> io_{nullptr};
+
+    uint32_t code_line_size_{0};
```

Comment:
when does this variable will be assigned a value

Reply:
https://github.com/LHT129/vsag/blob/graph_datacell/src/data_cell/graph_datacell.h#L98

Reply:
add a const modifier

Reply:
it can be changed if trans memory to disk